File: sagco-os/sagco/lang/lexer.py
"""
SAGCO Language Lexer.
Handles: strings (single/double quoted), numbers (int/float), cell refs (A1-ZZ999),
identifiers/keywords, all operators, #comments, newlines.
Returns list[Token].
"""
import re
from .tokens import Token, TokenType, KEYWORDS


class LexError(SyntaxError):
    pass


class Lexer:
    def __init__(self, source: str):
        self.source = source
        self.pos = 0
        self.line = 1
        self.col = 1
        self.tokens: list = []

    def error(self, msg: str):
        raise LexError(f"Lex error at line {self.line}, col {self.col}: {msg}")

    def peek(self, offset: int = 0) -> str:
        idx = self.pos + offset
        if idx < len(self.source):
            return self.source[idx]
        return ""

    def advance(self) -> str:
        ch = self.source[self.pos]
        self.pos += 1
        if ch == "\n":
            self.line += 1
            self.col = 1
        else:
            self.col += 1
        return ch

    def match(self, ch: str) -> bool:
        if self.pos < len(self.source) and self.source[self.pos] == ch:
            self.advance()
            return True
        return False

    def tok(self, ttype: TokenType, value: str, line: int, col: int) -> Token:
        return Token(ttype, value, line, col)

    def skip_whitespace(self):
        while self.pos < len(self.source) and self.source[self.pos] in (" ", "\t", "\r"):
            self.advance()

    def read_string(self, quote: str) -> Token:
        line, col = self.line, self.col
        self.advance()  # consume opening quote
        buf = []
        while self.pos < len(self.source):
            ch = self.source[self.pos]
            if ch == quote:
                self.advance()
                break
            if ch == "\\" and self.pos + 1 < len(self.source):
                self.advance()
                esc = self.advance()
                escape_map = {"n": "\n", "t": "\t", "r": "\r", "\\": "\\",
                              "'": "'", '"': '"'}
                buf.append(escape_map.get(esc, esc))
            elif ch == "\n":
                break  # unterminated string — just stop
            else:
                buf.append(self.advance())
        return self.tok(TokenType.STRING, "".join(buf), line, col)

    def read_number(self) -> Token:
        line, col = self.line, self.col
        buf = []
        if self.peek() == "-":
            buf.append(self.advance())
        while self.pos < len(self.source) and self.source[self.pos].isdigit():
            buf.append(self.advance())
        if self.pos < len(self.source) and self.source[self.pos] == ".":
            buf.append(self.advance())
            while self.pos < len(self.source) and self.source[self.pos].isdigit():
                buf.append(self.advance())
        return self.tok(TokenType.NUMBER, "".join(buf), line, col)

    def read_ident_or_keyword(self) -> Token:
        line, col = self.line, self.col
        buf = []
        while self.pos < len(self.source) and (self.source[self.pos].isalnum() or self.source[self.pos] == "_"):
            buf.append(self.advance())
        word = "".join(buf)
        upper = word.upper()

        # Check if it's a cell reference: letter(s) followed by digits
        cell_ref_match = re.fullmatch(r"[A-Za-z]{1,2}\d{1,3}", word)
        if cell_ref_match and upper not in KEYWORDS:
            return self.tok(TokenType.CELLREF, upper, line, col)

        if upper in KEYWORDS:
            ttype = KEYWORDS[upper]
            val = upper if ttype == TokenType.BOOL else word
            return self.tok(ttype, val, line, col)

        return self.tok(TokenType.IDENT, word, line, col)

    def read_comment(self) -> Token:
        line, col = self.line, self.col
        buf = []
        while self.pos < len(self.source) and self.source[self.pos] != "\n":
            buf.append(self.advance())
        return self.tok(TokenType.COMMENT, "".join(buf), line, col)

    def tokenize(self) -> list:
        """Tokenize the source. Returns list of Token (excludes COMMENT tokens)."""
        tokens = []
        while self.pos < len(self.source):
            self.skip_whitespace()
            if self.pos >= len(self.source):
                break
            ch = self.source[self.pos]
            line, col = self.line, self.col

            if ch == "\n":
                self.advance()
                # Emit NEWLINE but collapse multiples
                if not tokens or tokens[-1].type != TokenType.NEWLINE:
                    tokens.append(self.tok(TokenType.NEWLINE, "\n", line, col))
                continue

            if ch == "#":
                self.read_comment()  # discard
                continue

            if ch in ('"', "'"):
                tokens.append(self.read_string(ch))
                continue

            if ch.isdigit() or (ch == "-" and self.peek(1).isdigit() and
                                  (not tokens or tokens[-1].type in (
                                      TokenType.ASSIGN, TokenType.LPAREN,
                                      TokenType.COMMA, TokenType.PLUS,
                                      TokenType.MINUS, TokenType.NEWLINE,
                                      TokenType.EOF))):
                tokens.append(self.read_number())
                continue

            if ch.isalpha() or ch == "_":
                tokens.append(self.read_ident_or_keyword())
                continue

            # Operators
            self.advance()
            if ch == "=":
                if self.match("="):
                    tokens.append(self.tok(TokenType.EQ, "==", line, col))
                else:
                    tokens.append(self.tok(TokenType.ASSIGN, "=", line, col))
            elif ch == "!":
                if self.match("="):
                    tokens.append(self.tok(TokenType.NEQ, "!=", line, col))
                else:
                    self.error(f"Unexpected character '!'")
            elif ch == "<":
                if self.match("="):
                    tokens.append(self.tok(TokenType.LTEQ, "<=", line, col))
                else:
                    tokens.append(self.tok(TokenType.LT, "<", line, col))
            elif ch == ">":
                if self.match("="):
                    tokens.append(self.tok(TokenType.GTEQ, ">=", line, col))
                else:
                    tokens.append(self.tok(TokenType.GT, ">", line, col))
            elif ch == "+":
                tokens.append(self.tok(TokenType.PLUS, "+", line, col))
            elif ch == "-":
                tokens.append(self.tok(TokenType.MINUS, "-", line, col))
            elif ch == "*":
                tokens.append(self.tok(TokenType.STAR, "*", line, col))
            elif ch == "/":
                tokens.append(self.tok(TokenType.SLASH, "/", line, col))
            elif ch == "(":
                tokens.append(self.tok(TokenType.LPAREN, "(", line, col))
            elif ch == ")":
                tokens.append(self.tok(TokenType.RPAREN, ")", line, col))
            elif ch == "[":
                tokens.append(self.tok(TokenType.LBRACKET, "[", line, col))
            elif ch == "]":
                tokens.append(self.tok(TokenType.RBRACKET, "]", line, col))
            elif ch == ",":
                tokens.append(self.tok(TokenType.COMMA, ",", line, col))
            elif ch == ":":
                tokens.append(self.tok(TokenType.COLON, ":", line, col))
            elif ch == "$":
                # treat as start of a price string token
                buf = ["$"]
                while self.pos < len(self.source) and (self.source[self.pos].isdigit() or self.source[self.pos] == "."):
                    buf.append(self.advance())
                tokens.append(self.tok(TokenType.STRING, "".join(buf), line, col))
            else:
                # Unknown char — skip silently
                pass

        tokens.append(self.tok(TokenType.EOF, "", self.line, self.col))
        return tokens

File: BOARD-36-SAGCO-FORMULA-COMPILER/src/lexer.rs
use crate::token::{Span, Spanned, Token};
use thiserror::Error;

#[derive(Debug, Error)]
pub enum LexError {
    #[error("Unexpected character '{0}' at line {1}, col {2}")]
    UnexpectedChar(char, usize, usize),
    #[error("Unterminated string at line {0}, col {1}")]
    UnterminatedString(usize, usize),
}

pub struct Lexer {
    input: Vec<char>,
    pos: usize,
    line: usize,
    col: usize,
}

impl Lexer {
    pub fn new(input: &str) -> Self {
        // Strip leading '=' if present
        let s = if input.starts_with('=') {
            &input[1..]
        } else {
            input
        };
        Lexer {
            input: s.chars().collect(),
            pos: 0,
            line: 1,
            col: 1,
        }
    }

    fn peek(&self) -> Option<char> {
        self.input.get(self.pos).copied()
    }

    fn advance(&mut self) -> Option<char> {
        let c = self.input.get(self.pos).copied();
        self.pos += 1;
        if c == Some('\n') {
            self.line += 1;
            self.col = 1;
        } else {
            self.col += 1;
        }
        c
    }

    fn span(&self) -> Span {
        Span::new(self.line, self.col)
    }

    fn skip_whitespace(&mut self) {
        while matches!(self.peek(), Some(' ') | Some('\t') | Some('\r') | Some('\n')) {
            self.advance();
        }
    }

    fn read_number(&mut self) -> Token {
        let mut s = String::new();
        while let Some(c) = self.peek() {
            if c.is_ascii_digit() || c == '.' {
                s.push(c);
                self.advance();
            } else {
                break;
            }
        }
        Token::Number(s.parse().unwrap_or(0.0))
    }

    fn read_string(&mut self) -> Result<Token, LexError> {
        let (line, col) = (self.line, self.col);
        self.advance(); // consume opening quote
        let mut s = String::new();
        loop {
            match self.peek() {
                Some('"') => {
                    self.advance();
                    return Ok(Token::Str(s));
                }
                Some(c) => {
                    s.push(c);
                    self.advance();
                }
                None => return Err(LexError::UnterminatedString(line, col)),
            }
        }
    }

    /// Read an identifier or cell ref (e.g. A1, B2, SUM, IF)
    fn read_ident_or_cell(&mut self) -> Token {
        let mut s = String::new();
        while let Some(c) = self.peek() {
            if c.is_alphanumeric() || c == '_' {
                s.push(c);
                self.advance();
            } else {
                break;
            }
        }

        // Check if it's a range ref like A1:B10 — peek ahead
        // Already consumed ident; check if next is ':' and next-next looks like cell
        if self.peek() == Some(':') {
            // peek to see if right side is a cell ref
            let saved_pos = self.pos;
            let saved_line = self.line;
            let saved_col = self.col;
            self.advance(); // consume ':'
            let mut right = String::new();
            while let Some(c) = self.peek() {
                if c.is_alphanumeric() {
                    right.push(c);
                    self.advance();
                } else {
                    break;
                }
            }
            if is_cell_ref(&right) {
                return Token::RangeRef(s, right);
            } else {
                // Not a range, backtrack
                self.pos = saved_pos;
                self.line = saved_line;
                self.col = saved_col;
            }
        }

        // Check keywords
        match s.as_str() {
            "TRUE" | "true" => return Token::Bool(true),
            "FALSE" | "false" => return Token::Bool(false),
            _ => {}
        }

        if is_cell_ref(&s) {
            Token::CellRef(s)
        } else {
            Token::Ident(s)
        }
    }

    pub fn tokenize(&mut self) -> Result<Vec<Spanned<Token>>, LexError> {
        let mut tokens = Vec::new();
        loop {
            self.skip_whitespace();
            let span = self.span();
            match self.peek() {
                None => {
                    tokens.push(Spanned::new(Token::EOF, span));
                    break;
                }
                Some(c) => {
                    let tok = match c {
                        '(' => { self.advance(); Token::LParen }
                        ')' => { self.advance(); Token::RParen }
                        ',' => { self.advance(); Token::Comma }
                        ';' => { self.advance(); Token::Semicolon }
                        '+' => { self.advance(); Token::Plus }
                        '-' => { self.advance(); Token::Minus }
                        '*' => { self.advance(); Token::Star }
                        '/' => { self.advance(); Token::Slash }
                        '^' => { self.advance(); Token::Caret }
                        '&' => { self.advance(); Token::Ampersand }
                        '<' => {
                            self.advance();
                            if self.peek() == Some('=') {
                                self.advance();
                                Token::LtEq
                            } else if self.peek() == Some('>') {
                                self.advance();
                                Token::NEq
                            } else {
                                Token::Lt
                            }
                        }
                        '>' => {
                            self.advance();
                            if self.peek() == Some('=') {
                                self.advance();
                                Token::GtEq
                            } else {
                                Token::Gt
                            }
                        }
                        '=' => {
                            self.advance();
                            if self.peek() == Some('=') {
                                self.advance();
                            }
                            Token::Eq
                        }
                        '!' => {
                            self.advance();
                            if self.peek() == Some('=') {
                                self.advance();
                                Token::NEq
                            } else {
                                Token::Bang
                            }
                        }
                        ':' => { self.advance(); Token::Colon }
                        '"' => self.read_string()?,
                        c if c.is_ascii_digit() || c == '.' => self.read_number(),
                        c if c.is_alphabetic() || c == '_' => self.read_ident_or_cell(),
                        other => {
                            return Err(LexError::UnexpectedChar(other, self.line, self.col));
                        }
                    };
                    tokens.push(Spanned::new(tok, span));
                }
            }
        }
        Ok(tokens)
    }
}

/// A cell ref has uppercase letters followed by digits, e.g. A1, B10, AB3
fn is_cell_ref(s: &str) -> bool {
    if s.is_empty() {
        return false;
    }
    let mut chars = s.chars().peekable();
    let mut has_letters = false;
    let mut has_digits = false;
    while let Some(&c) = chars.peek() {
        if c.is_ascii_uppercase() {
            has_letters = true;
            chars.next();
        } else {
            break;
        }
    }
    while let Some(&c) = chars.peek() {
        if c.is_ascii_digit() {
            has_digits = true;
            chars.next();
        } else {
            break;
        }
    }
    has_letters && has_digits && chars.peek().is_none()
}

pub fn tokenize(input: &str) -> Result<Vec<Spanned<Token>>, LexError> {
    Lexer::new(input).tokenize()
}
